File: typescript/src/session.ts
import { alloc, concat, createFrom, fromInteger } from 'stedy/bytes'
import {
  CertificationResult,
  CertifyFunction,
  DecryptFunction,
  EncryptFunction,
  SessionFunction,
  SignFunction,
  VerifyFunction
} from '../types'
import { decrypt, encrypt } from './crypto/cipher'
import { verify as verifySignature } from './crypto/sign'
import { createErrorSignResult, ensureSignResult } from './utils'

const verifySession = async (
  transcript: BufferSource,
  theirIdentityKey: BufferSource,
  theirSecretKey: BufferSource,
  message: BufferSource
) => {
  try {
    const signature = await decrypt(theirSecretKey, 0, message)
    const verified = await verifySignature(
      transcript,
      theirIdentityKey,
      signature
    )
    return verified
  } catch (error) {
    return false
  }
}

const createCertify =
  (sign: SignFunction, theirPublicKey: BufferSource): CertifyFunction =>
  async (data?: BufferSource) => {
    try {
      const result = await sign(concat([data, theirPublicKey]))
      return ensureSignResult(result) as CertificationResult
    } catch (error) {
      return createErrorSignResult() as CertificationResult
    }
  }

const createDecrypt =
  (theirSecretKey: BufferSource): DecryptFunction =>
  async (message: BufferSource) => {
    const [nonce, ciphertext] = createFrom(message).read(4)
    try {
      const data = await decrypt(
        theirSecretKey,
        nonce.readUint32BE(),
        ciphertext
      )
      return { success: true, data }
    } catch (error) {
      return {
        success: false,
        data: alloc(Math.max(ciphertext.byteLength - 16, 0))
      }
    }
  }

const createEncrypt = (ourSecretKey: BufferSource): EncryptFunction => {
  let index = 0
  return async (data: BufferSource) => {
    index += 1
    try {
      const ciphertext = await encrypt(ourSecretKey, index, data)
      const nonce = fromInteger(index)
      const message = concat([nonce, ciphertext])
      return { success: true, message }
    } catch (error) {
      return { success: false, message: alloc(data.byteLength + 16) }
    }
  }
}

const createVerify =
  (theirIdentityKey: BufferSource): VerifyFunction =>
  async (certificates: BufferSource, data?: BufferSource) => {
    try {
      const subject = concat([data, theirIdentityKey])
      const results = await Promise.all(
        createFrom(certificates)
          .split(96)
          .map((certificate) => {
            const [identityKey, signature] = certificate.read(32)
            return verifySignature(subject, identityKey, signature)
          })
      )
      return results.length > 0 && results.every((result) => result === true)
    } catch (error) {
      return false
    }
  }

const createSession =
  (
    sign: SignFunction,
    theirIdentityKey: BufferSource,
    transcript: BufferSource,
    ourSecretKey: BufferSource,
    theirSecretKey: BufferSource
  ): SessionFunction =>
  async (message: BufferSource) => {
    const success = await verifySession(
      transcript,
      theirIdentityKey,
      theirSecretKey,
      message
    )
    const certify = createCertify(sign, theirIdentityKey)
    const decrypt = createDecrypt(theirSecretKey)
    const encrypt = createEncrypt(ourSecretKey)
    const verify = createVerify(theirIdentityKey)
    const session = {
      certify,
      decrypt,
      encrypt,
      verify
    }
    return { success, session }
  }

export default createSession

File: cplusplus/src/session.cpp
#include "session.h"

#include "numbers.h"
#include "private.h"
#include "sizes.h"

namespace Autograph {

DecryptFunction createDecrypt(Bytes theirSecretKey) {
  Bytes messageIndex(INDEX_SIZE);
  Bytes decryptIndex(INDEX_SIZE);
  Bytes plaintextSize(SIZE_SIZE);
  Bytes skippedKeys(SKIPPED_KEYS_SIZE);
  auto decryptFunction = [theirSecretKey, messageIndex, decryptIndex,
                          skippedKeys,
                          plaintextSize](const Bytes message) mutable {
    Bytes plaintext(getPlaintextSize(message.size()));
    bool success = autograph_decrypt(plaintext.data(), plaintextSize.data(),
                                     messageIndex.data(), decryptIndex.data(),
                                     skippedKeys.data(), theirSecretKey.data(),
                                     message.data(), message.size()) == 0;
    if (success) {
      plaintext.resize(autograph_read_uint32(plaintextSize.data()));
    }
    DecryptionResult result = {
        success, autograph_read_uint64(messageIndex.data()), plaintext};
    return result;
  };
  return decryptFunction;
}

EncryptFunction createEncrypt(Bytes ourSecretKey) {
  Bytes index(INDEX_SIZE);
  auto encryptFunction = [ourSecretKey, index](const Bytes plaintext) mutable {
    Bytes ciphertext(getCiphertextSize(plaintext.size()));
    bool success =
        autograph_encrypt(ciphertext.data(), index.data(), ourSecretKey.data(),
                          plaintext.data(), plaintext.size()) == 0;
    EncryptionResult result = {success, autograph_read_uint64(index.data()),
                               ciphertext};
    return result;
  };
  return encryptFunction;
}

SignDataFunction createSignData(const SignFunction sign,
                                const Bytes theirPublicKey) {
  auto signDataFunction = [sign, theirPublicKey](const Bytes data) {
    Bytes subject(getSubjectSize(data.size()));
    autograph_subject(subject.data(), theirPublicKey.data(), data.data(),
                      data.size());
    auto signResult = sign(subject);
    SignResult result = {signResult.success, signResult.signature};
    return result;
  };
  return signDataFunction;
}

SignIdentityFunction createSignIdentity(const SignFunction sign,
                                        const Bytes theirPublicKey) {
  auto signIdentityFunction = [sign, theirPublicKey]() {
    auto signResult = sign(theirPublicKey);
    SignResult result = {signResult.success, signResult.signature};
    return result;
  };
  return signIdentityFunction;
}

unsigned int countCertificates(const Bytes certificates) {
  return certificates.size() / (PUBLIC_KEY_SIZE + SIGNATURE_SIZE);
}

VerifyDataFunction createVerifyData(const Bytes theirIdentityKey) {
  auto verifyDataFunction = [theirIdentityKey](const Bytes certificates,
                                               const Bytes data) {
    return autograph_verify_data(theirIdentityKey.data(), certificates.data(),
                                 countCertificates(certificates), data.data(),
                                 data.size()) == 0;
  };
  return verifyDataFunction;
}

VerifyIdentityFunction createVerifyIdentity(const Bytes theirIdentityKey) {
  auto verifyIdentityFunction = [theirIdentityKey](const Bytes certificates) {
    return autograph_verify_identity(theirIdentityKey.data(),
                                     certificates.data(),
                                     countCertificates(certificates)) == 0;
  };
  return verifyIdentityFunction;
}

}  // namespace Autograph
